ding/policy/qgpo.py

```diff
@@ -0,0 +1,172 @@
+class QGPOPolicy(Policy):
+    """
+       Overview:
+            Policy class of QGPO algorithm
+            Contrastive Energy Prediction for Exact Energy-Guided Diffusion Sampling in Offline Reinforcement Learning
+            https://arxiv.org/abs/2304.12824
+    """
+
+    config = dict(
+        # (str) RL policy register name (refer to function "POLICY_REGISTRY").
+        type='qgpo',
+        # (bool) Whether to use cuda for network.
+        cuda=False,
+        # (bool type) on_policy: Determine whether on-policy or off-policy.
+        # on-policy setting influences the behaviour of buffer.
+        # Default False in QGPO.
+        on_policy=False,
+        multi_agent=False,
+        model=dict(
+            score_net=dict(
+                qgpo_critic=dict(
+                    # (float) The scale of the energy guidance when training qt.
+                    # \pi_{behavior}\exp(f(s,a)) \propto \pi_{behavior}\exp(alpha * Q(s,a))
+                    alpha=3,
+                    # (float) The scale of the energy guidance when training q0.
+                    # \mathcal{T}Q(s,a)=r(s,a)+\mathbb{E}_{s'\sim P(s'|s,a),a'\sim\pi_{support}(a'|s')}Q(s',a')
+                    # \pi_{support} \propto \pi_{behavior}\exp(q_alpha * Q(s,a))
+                    q_alpha=1,
+                ),
+            ),
+            device='cuda',
+            # obs_dim
+            # action_dim
+        ),
+        learn=dict(
+            # learning rate for behavior model training
+            learning_rate=1e-4,
+            # batch size during the training of behavior model
+            batch_size=4096,
+            # batch size during the training of q value
+            batch_size_q=256,
+            # number of fake action support
+            M=16,
+            # number of diffusion time steps
+            diffusion_steps=15,
+            # training iterations when behavior model is fixed
+            behavior_policy_stop_training_iter=600000,
+            # training iterations when energy-guided policy begin training
+            energy_guided_policy_begin_training_iter=600000,
+            # training iterations when q value stop training, default None means no limit
+            q_value_stop_training_iter=1100000,
+        ),
+        eval=dict(
+            # energy guidance scale for policy in evaluation
+            # \pi_{evaluation} \propto \pi_{behavior}\exp(guidance_scale * alpha * Q(s,a))
+            guidance_scale=[0.0, 1.0, 2.0, 3.0, 5.0, 8.0, 10.0],
+        ),
+    )
+
+    def _init_learn(self) -> None:
+        self.cuda = self._cfg.cuda
+        if self.cuda:
+            self.margin_prob_std_fn = functools.partial(marginal_prob_std, device=self._device)
+        self.behavior_model_optimizer = torch.optim.Adam(
+            self._model.score_model.parameters(), lr=self._cfg.learn.learning_rate
+        )
+        self.behavior_policy_stop_training_iter = self._cfg.learn.behavior_policy_stop_training_iter if hasattr(
```

Comment:
why do we use if expression here, we have already set this field in default config

Reply:
fixed